tsinfer/formats.py

```diff
@@ -2735,7 +2755,9 @@ def num_sites(self):
         """
         The number of inference sites used to generate the ancestors
         """
-        return self.sites_position.shape[0]
+        if not hasattr(self, "_num_sites"):
```

Comment:
Using hasattr feels a bit icky (and slow I think?). How about initialising ``self._num_sites = None`` and using that as the caching tag?

Reply:
I wrote it that way first - but `load` calls `__new__` and not `__init__`. The best way is probably a caching decorator. Will see if our python version supports using the one in functools.

Reply:
Woop! Now that we have dropped 3.7 we can use `@functools.cached_property`!